peakina/readers/excel.py

```diff
@@ -0,0 +1,199 @@
+    skiprows: Optional[int] = None,
+    nrows: Optional[int] = None,
+) -> List[str]:
+
+    row_iterator = _get_rows_iterator(wb, excel_type, sh_name, preview)
+
+    if excel_type is EXCEL_TYPE.NEW:
+        for row_number, gen in enumerate(row_iterator):
+            for row in gen:
+                if skiprows:
+                    if row_number < skiprows:
+                        continue
+                row_number, row_subset = _build_row_subset(
+                    row, sh_name, sheetnames, row_number, row_subset
+                )
+                if nrows:
+                    if row_number == nrows:
+                        break
+    else:
+        for row_number, row in enumerate(row_iterator):
+            if skiprows:
+                if row_number < skiprows:
+                    continue
+            row_number, row_subset = _build_row_subset(
+                row, sh_name, sheetnames, row_number, row_subset
+            )
+            if nrows:
+                if row_number == nrows:
+                    break
+    return row_subset
+
+
+def _read_sheets(
+    wb: Any,
+    excel_type: EXCEL_TYPE,
+    sheet_names: List[Any],
+    preview: Optional[PreviewArgs],
+    nrows: Optional[int] = None,
+    skiprows: Optional[int] = None,
+) -> List[Any]:
+    """
+    This method will loop over sheets, read content and return a list of rows
+    depending on your inputs
+
+    """
+
+    row_subset: List[str] = []
+    for sh_name in sheet_names:
+        row_subset = _get_row_subset_per_sheet(
+            wb, sh_name, sheet_names, preview, excel_type, row_subset, skiprows, nrows
+        )
+
+    if excel_type is EXCEL_TYPE.NEW:
+        wb.close()
+
+    # cleaning extras rows with __sheet__
+    if len(sheet_names) > 1:
+        row_subset[1:] = [x for x in row_subset[1:] if "__sheet__" not in x]
+
+    return row_subset
+
+
+def read_excel(
+    filepath: str,
+    *,
+    preview: Optional[PreviewArgs] = None,
+    sheet_name: str = "",
+    na_values: Any = None,
+    keep_default_na: bool = False,
+    skiprows: Optional[int] = None,
+    nrows: Optional[int] = None,
+) -> pd.DataFrame:
+    """
+    The read_excel function is using openpyxl to parse the csv file and read it
+
+    """
+
+    column_names = []
+
+    try:
+        excel_type, wb = EXCEL_TYPE.NEW, load_workbook(filepath, read_only=True)
+        all_sheet_names = wb.sheetnames
+
+        if preview:
+            # we get column names with the iterator
+            for sh_name in all_sheet_names:
+                column_names += [c.value for c in next(wb[sh_name].iter_rows(min_row=1, max_row=1))]
+
+    except InvalidFileException:
+        excel_type, wb = EXCEL_TYPE.OLD, xlrd.open_workbook(filepath)
+        all_sheet_names = wb.sheet_names()
+
+        if preview:
+            for sh_name in all_sheet_names:
+                column_names += [c.value for c in wb[sh_name].row(0)]
+
+    sheet_names = [sheet_name] if sheet_name else all_sheet_names
```

Comment:
:x: I think this is a breaking change, which may affect how data is extracted from excel in apps until now.

Today we're using [pd.read_excel[https://pandas.pydata.org/docs/reference/api/pandas.read_excel.html). In this method, providing no sheet name means "extracting the first sheet".
If I read correctly, with this mew method `read_excel`, providing no sheet names would mean concatenating all sheets together.
I think this is not what we want.

Reply:
Thanks for your comment !
We were having a previous test that was already concatenating sheets when the parameter "sheet_name" was None https://github.com/ToucanToco/peakina/blob/main/tests/test_datasource.py#L156 and then add a new column `__sheet__` for corresponding sheets 🤔
```python
def test_multi_sheets_excel(path):
    """It should add a __sheet__ column when retrieving multiple sheet"""
    ds = DataSource(path("fixture-multi-sheet.xlsx"), reader_kwargs={"sheet_name": None})
    df = pd.DataFrame({"Month": [1, 2], "Year": [2019, 2019], "__sheet__": ["January", "February"]})
    assert ds.get_df().equals(df)
```
don't know well how it could break our current extraction...

Reply:
Hm...

I just did the test in our test instance, and I do reproduce the behavior when the first sheet is selected  with the following:
[Two Sheets.xlsx](https://github.com/ToucanToco/peakina/files/8069595/Two.Sheets.xlsx)
![Screenshot from 2022-02-15 14-46-31](https://user-images.githubusercontent.com/932583/154074965-be1f36a0-0491-4430-88e4-94310d6ab537.png)
![Screenshot from 2022-02-15 14-48-01](https://user-images.githubusercontent.com/932583/154074992-487fa13c-5d3b-4a23-9915-8d7872e83b89.png)
![Screenshot from 2022-02-15 14-48-11](https://user-images.githubusercontent.com/932583/154075010-6e56c4f3-adbe-4dd9-ae03-fc1c28e7fe70.png)
 
What I noticed is that, since the new UI to configure excel files, it's not possible any more to have `sheet: undefined` in the DATA_SOURCE configuration. Yet it still possible to write it in code mode, which makes me think it could still break old apps for which the configuration has not been updated.

(I didn't find any migration in the back-end that would take of this, only a "migration" in the front-end, so only when the data-source is edited).

Maybe there is another place in the back-end where we select the first sheet (before peakina), but I'd be more confident to merge this if we can explain this behavior :)
(Sorry for being cautious!)